src/main/java/com/amazon/ion/impl/MarkerList.java

```diff
@@ -0,0 +1,103 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.ion.impl;
+
+/**
+ * A list of {@link Marker} values that grows as necessary, and serves as a pool to avoid excessive allocation.
+ */
+public class MarkerList {
+
+    private Marker[] data;
+    private int numberOfValues;
+    private int provisionalIndex;
+
+    /**
+     * Constructs a new MarkerList with the specified capacity.
+     * @param initialCapacity   The number of Markers that can be stored in this MarkerList before it will need to be
+     *                          reallocated.
+     */
+    public MarkerList(final int initialCapacity) {
+        data = new Marker[initialCapacity];
+        for (int i = 0; i < initialCapacity; i++) {
+            data[i] = new Marker(-1, -1);
+        }
+        numberOfValues = 0;
+        provisionalIndex = 0;
+    }
+
+    /**
+     * Accessor.
+     * @return  The number of Markers currently stored in the list.
+     */
+    public int size() {
+        return numberOfValues;
+    }
+
+    /**
+     * @return {@code true} if there are Markers stored in the list.
+     */
+    public boolean isEmpty() {
+        return numberOfValues == 0;
+    }
+
+    /**
+     * Empties the list.
+     * Note that this method does not shrink the size of the backing data store.
+     */
+    public void clear() {
+        numberOfValues = 0;
+        provisionalIndex = 0;
+    }
+
+    /**
+     * Returns the {@code index}th Marker in the list.
+     * @param index     The list index of the desired Marker.
+     * @return          The Marker at index {@code index} in the list.
+     * @throws IndexOutOfBoundsException    if the index is negative or greater than the number of Markers stored in the
+     *                                      list.
+     */
+    public Marker get(int index) {
+        if (index < 0 || index >= numberOfValues) {
+            throw new IndexOutOfBoundsException(
+                "Invalid index " + index + " requested from IntList with " + numberOfValues + " values."
+            );
+        }
+        return data[index];
+    }
+
+    /**
+     * @return The Marker that, if committed, will become the next element in the list. Grows the list if necessary.
+     */
+    public Marker provisionalElement() {
```

Comment:
I don't understand why there needs to be provisional elements. I don't recall seeing any cases where a provisional element was created but not committed. Can you clarify the purpose of this?

Reply:
See [IonCursorBinary.slowReadAnnotationWrapperHeader_1_1](https://github.com/amazon-ion/ion-java/pull/666/files#diff-b003a36a163c1b1b468b1080c3d0fec389784bd6938eb9a5dcf6c1385935e02fR1080). There are early returns before commit in cases where EOF is encountered before completing the annotations. The continuable reader needs to be able to pick up where it left off.